pkg/container/versioned/value.go

```diff
@@ -0,0 +1,440 @@
+)
+
+type version uint64
+
+const (
+	// invalidVersion is never found from the value range.
+	// Also used as the upper bound for new values, so that the value is
+	// found when looking with 'maxVersion'
+	invalidVersion = version(math.MaxUint64)
+
+	// maxVersion in a VersionHold finds the latest version of all non-removed values
+	maxVersion = version(math.MaxUint64 - 1)
+)
+
+// KeepVersion is an exported version type used when releasing memory held for old versions.
+type KeepVersion version
+
+// NextVersion is an exported version type used to add or remove new versions of values.
+type NextVersion version
+
+var (
+	errInvalidVersion  = errors.New("invalid version")
+	errStaleVersion    = errors.New("stale version")
+	errVersionNotFound = errors.New("version not found")
+)
+
+// VersionHold is used to keep values valid for a specific version from being released, so that they
+// are available for use for as long as the VersionHold is not closed.
+//
+// A special form with a nil coordinator, which is returned by Latest(), always finds the latest
+// versions and does not keep any versions from being released.
+type VersionHold struct {
+	// 'version' is the version this VersionHold keeps from being released
+	version version
+	// Coordinator of this VersionHold, if any. Used for closing if non-nil.
+	// Atomic due to nilling and for copy prevention.
+	coordinator atomic.Pointer[Coordinator]
+}
+
+func (h *VersionHold) String() string {
+	if h == nil {
+		return "version: <nil>"
+	}
+	return fmt.Sprintf("version: %d", h.version)
+}
+
+// versionHoldFinalizer is used to warn about missing Close() calls.
+func versionHoldFinalizer(h *VersionHold) {
+	slog.Error("Hold for SelectorCache version not closed.", logfields.Version, h.version, logfields.LogSubsys, "containers/versioned")
+	h.Close()
+}
+
+func newVersionHold(version version, coordinator *Coordinator) *VersionHold {
+	// hold on maxVersion never expires
+	if version == maxVersion {
+		coordinator = nil
+	}
+	h := &VersionHold{version: version}
+	h.coordinator.Store(coordinator)
+	if coordinator != nil {
+		// Set a finalizer to catch unclosed holds. The finalizer function complains
+		// loudly, so that we do not rely the finalizer for closing.
+		runtime.SetFinalizer(h, versionHoldFinalizer)
+	}
+	return h
+}
+
+// Latest returns a VersionHold for the latest version of current/non-removed values
+// Only to be used in cases where the latest values are needed and no transactionality is required.
+func Latest() *VersionHold {
+	return newVersionHold(maxVersion, nil)
+}
+
+func (h *VersionHold) IsValid() bool {
+	return h != nil && (h.version == maxVersion || h.coordinator.Load() != nil)
+}
+
+// Close releases the held version for removal once no hold for it are no longer held.
+// This may not be called while holding any locks that the 'closer' function passed to the
+// coordinator may take!
+func (h *VersionHold) Close() (err error) {
```

Comment:
Idea: in debug mode, add a finalizer that panics if Close() was not called before the value was GCd.

Reply:
I like this, but do we want the panic to happen if a customer runs Cilium in debug mode? The warning log is there to draw attention to this, and the stacktrace provides the same information a panic would.